sound/soc/sof/intel/hda-dsp.c

```diff
@@ -767,3 +771,33 @@ int hda_dsp_set_hw_params_upon_resume(struct snd_sof_dev *sdev)
 #endif
 	return 0;
 }
+
+void hda_dsp_d0i3_work(struct work_struct *work)
+{
+	struct sof_intel_hda_dev *hdev = container_of(work,
+						      struct sof_intel_hda_dev,
+						      d0i3_work.work);
+	struct hdac_bus *bus = &hdev->hbus.core;
+	struct snd_sof_dev *sdev = dev_get_drvdata(bus->dev);
+	struct sof_dsp_power_state target_state;
+	int ret;
+
+	target_state.state = SOF_DSP_D0;
+
+	/* DSP can enter D0I3 iff only D0I3-compatible streams are active */
+	if (snd_sof_dsp_only_d0i3_streams_active(sdev))
+		target_state.substate = SOF_HDA_DSP_D0I3;
+	else
+		target_state.substate = SOF_HDA_DSP_D0I0;
+
+	/* remain in D0I0 */
+	if (target_state.substate == SOF_HDA_DSP_D0I0)
+		return;
+
+	/* This can fail but error cannot be propagated */
+	ret = hda_dsp_set_power_state(sdev, &target_state);
+	if (ret < 0)
+		dev_err_ratelimited(sdev->dev,
+				    "error: failed to set DSP state %d substate %d\n",
+				    target_state.state, target_state.substate);
+}
```

Comment:
if this fails, did the state change or not?

In other words is this 'atomic' with a state change on success of all operations? or could we be left with partial state on both sides?

Reply:
@plbossart yes we could have a case where the driver switched the state but the PM_GATE ipc failed and we'll left in a partial state. Let me fix this part. Thanks for spotting this